crates/kubelet/src/pod/queue.rs

```diff
@@ -19,79 +30,166 @@ use crate::provider::Provider;
+
+                    let check = async {
+                        loop {
+                            tokio::time::delay_for(std::time::Duration::from_secs(1)).await;
+                            if *check_pod_deleted.read().await {
+                                break;
+                            }
+                        }
+                    };
+
+                    tokio::select! {
+                        result = run_to_completion(&task_client, state, &mut pod_state, &pod) => match result {
+                            Ok(()) => info!("Pod {} state machine exited without error", name),
+                            Err(e) => {
+                                error!("Pod {} state machine exited with error: {:?}", name, e);
+                                let api: kube::Api<KubePod> = kube::Api::namespaced(task_client.clone(), pod.namespace());
+                                let patch = serde_json::json!(
+                                    {
+                                        "metadata": {
+                                            "resourceVersion": "",
+                                        },
+                                        "status": {
+                                            "phase": Phase::Failed,
+                                            "reason": format!("{:?}", e),
+                                            "containerStatuses": Vec::<()>::new(),
+                                            "initContainerStatuses": Vec::<()>::new(),
+                                        }
+                                    }
+                                );
+                                let data = serde_json::to_vec(&patch).unwrap();
+                                api.patch_status(&pod.name(), &kube::api::PatchParams::default(), data)
+                                    .await.unwrap();
+                            },
+                        },
+                        _ = check => {
+                            let state: P::TerminatedState = Default::default();
+                            info!("Pod {} terminated. Jumping to state {:?}.", name, state);
+                            match run_to_completion(&task_client, state, &mut pod_state, &pod).await {
+                                Ok(()) => info!("Pod {} state machine exited without error", name),
+                                Err(e) => error!("Pod {} state machine exited with error: {:?}", name, e),
+                            }
+                        }
+                    }
+
+                    info!("Pod {} waiting for deregistration.", name);
+                    loop {
+                        tokio::time::delay_for(std::time::Duration::from_secs(1)).await;
+                        if *check_pod_deleted.read().await {
+                            info!("Pod {} deleted.", name);
+                            break;
+                        }
+                    }
+                    pod_state.async_drop().await;
```

Comment:
Nit: Any way we could combine these into a single `drop`?

Reply:
Maybe, I'm not convinced that AsyncDrop is really what we want to call this. It would be good to make sure that it is always dropped when async_drop is called.